include/log.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <ostream>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+#include <variant>
+
+// TODO: Make this configurable through CMake?
+#define SPDLOG_ACTIVE_LEVEL SPDLOG_LEVEL_TRACE
+#include <spdlog/spdlog.h>
+
+#include <spdlog/fmt/ostr.h>
+#include <spdlog/sinks/stdout_color_sinks.h>
+
+#include "print_utils.h"
+
+#define CELERITY_SET_LOG_CTX(ctx) CELERITY_DETAIL_SET_LOG_CTX(ctx)
+
+#define CELERITY_TRACE(...) SPDLOG_TRACE(*celerity::detail::active_log_ctx + __VA_ARGS__)
+#define CELERITY_DEBUG(...) SPDLOG_DEBUG(*celerity::detail::active_log_ctx + __VA_ARGS__)
+#define CELERITY_INFO(...) SPDLOG_INFO(*celerity::detail::active_log_ctx + __VA_ARGS__)
+#define CELERITY_WARN(...) SPDLOG_WARN(*celerity::detail::active_log_ctx + __VA_ARGS__)
+#define CELERITY_ERROR(...) SPDLOG_ERROR(*celerity::detail::active_log_ctx + __VA_ARGS__)
+#define CELERITY_CRITICAL(...) SPDLOG_CRITICAL(*celerity::detail::active_log_ctx + __VA_ARGS__)
+
+namespace celerity {
+namespace detail {
+
+	using log_map = std::unordered_map<std::string, std::variant<std::string_view, size_t>>;
+	using log_level = spdlog::level::level_enum;
+
+	struct log_context {
+		std::string value;
+		log_context() = default;
+		explicit log_context(const log_map& values) { value = fmt::format("[{}] ", values); }
+	};
+
+	inline const std::string null_log_ctx;
+	inline thread_local const std::string* active_log_ctx = &null_log_ctx;
+
+	struct log_ctx_setter {
+		log_ctx_setter(log_context& ctx) { celerity::detail::active_log_ctx = &ctx.value; }
+		~log_ctx_setter() { celerity::detail::active_log_ctx = &celerity::detail::null_log_ctx; }
+	};
+
+#define CELERITY_DETAIL_SET_LOG_CTX(ctx)                                                                                                                       \
+	log_ctx_setter _set_log_ctx { ctx }
+
+} // namespace detail
+} // namespace celerity
+
+template <>
+struct fmt::formatter<celerity::detail::log_map> {
+	constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+
+	template <typename FormatContext>
+	auto format(const celerity::detail::log_map& map, FormatContext& ctx) {
```

Comment:
Wouldn't it be easier to move this code into the `log_context` constructor instead? It appears to be the only use.

Reply:
It's also used in the `user_bench` utilities!

Reply:
Right!